### PR TITLE
Split code into modules based on the version of edoc/edek they're used for

### PR DESCRIPTION
**Pull request by coltfred**
This diff is especially messy because I broke code into so many different modules. I'll try to describe my logic so we can tell in the future why things are the way they are.

Things at the top level have to do with the V4DocumentHeader. Creating it, signing it, etc.

`v4` - Contains things to do with v4 specific things. This has a bunch of attached code in it currently, which isn't being used by ironcore-alloy, but a PR is incoming once the dust settles.
`v5` - `key_id_header`, `EncryptedPayload`.
`v3` - Specific things for v3 edoc and edek.

### CODE REVIEW

File: src/aes.rs [hunk 1]
```diff
@@ -1,158 +1,84 @@
+// This module is dedicated with things to do with oes encryption/decryption.
```
Comment (giarc3) [inline src/aes.rs]:
```suggestion
// This module is dedicated with things to do with aes encryption/decryption.
```

File: src/aes.rs [hunk 2]
```diff
@@ -1,158 +1,84 @@
...
-    let aes_edek = icl_header_v4::v4document_header::edek_wrapper::Aes256GcmEncryptedDek {
-        ciphertext: edek.0.into(),
-        iv: Bytes::copy_from_slice(&iv),
-        id: id.into(),
-        ..Default::default()
-    };
-    Ok((dek, aes_edek))
+impl Default for IvAndCiphertext {
+    fn default() -> IvAndCiphertext {
+        IvAndCiphertext([].as_ref().into())
```
Comment (giarc3) [inline src/aes.rs]:
```suggestion
        IvAndCiphertext(Bytes::new())
```

File: src/aes.rs [hunk 3]
```diff
@@ -1,158 +1,84 @@
...
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct EncryptionKey(pub [u8; 32]);
+
+/// Decrypt the AES encrypted payload using the key. Note that the IV is on the front of the payload and the tag
+/// is on the end.
+pub fn decrypt_document_with_attached_iv(
     key: &EncryptionKey,
-    attached_encrypted_payload: &[u8],
+    aes_encrypted_payload: &[u8],
 ) -> std::result::Result<PlaintextDocument, Error> {
```
Comment (giarc3) [inline src/aes.rs]:
```suggestion
) -> Result<PlaintextDocument> {
```

File: src/v3/mod.rs [hunk 4]
```diff
@@ -12,6 +14,8 @@ const GCM_TAG_LEN: usize = 16;
 
 const V3: u8 = 3u8;
 
+const MAGIC: &[u8; 4] = crate::v4::MAGIC;
```
Comment (giarc3) [inline src/v3/mod.rs]: How come this instead of the import?

File: src/aes.rs [hunk 5]
```diff
@@ -1,158 +1,84 @@
...
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct EncryptionKey(pub [u8; 32]);
+
+/// Decrypt the AES encrypted payload using the key. Note that the IV is on the front of the payload and the tag
+/// is on the end.
+pub fn decrypt_document_with_attached_iv(
     key: &EncryptionKey,
-    attached_encrypted_payload: &[u8],
+    aes_encrypted_payload: &[u8],
```
Comment (giarc3) [inline src/aes.rs]: Should this take IvAndCiphertext now?

File: src/v3/mod.rs [hunk 6]
```diff
@@ -67,7 +71,7 @@ impl EncryptedPayload {
     /// Decrypt a V3 detached document and verify its signature.
     pub fn decrypt(self, key: &EncryptionKey) -> Result<PlaintextDocument, Error> {
         if verify_signature(key.0, &self.v3_document_header) {
-            decrypt_attached_document_core(key, &self.iv_and_ciphertext)
+            decrypt_document_with_attached_iv(key, &self.iv_and_ciphertext)
```
Comment (giarc3) [inline src/v3/mod.rs:74]: (Can't comment the above line)
Should `self.iv_and_ciphertext` be an `IvAndCiphertext` now?

File: src/v4/attached.rs [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+fn get_v4_header_and_payload(mut b: Bytes) -> Result<(Bytes, IvAndCiphertext)> {
+    let initial_len = b.len();
+    if initial_len >= PRE_HEADER_LEN {
+        let first_byte = b.get_u8();
+        if first_byte == V4 {
+            //Check to see if the next 4 bytes are the IRON ascii chars
+            let maybe_magic = b.split_to(MAGIC.len());
+            Some(maybe_magic)
+                .filter(|bytes| *bytes == MAGIC[..])
```
Comment (giarc3) [inline src/v4/attached.rs]:
```suggestion
                .filter(|bytes| bytes.as_ref() == MAGIC)
```

File: src/v4/attached.rs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                .ok_or(Error::NoIronCoreMagic)?;
+            //The following 2 bytes should be a u16 (big endian). This is the size of the PB header
+            let header_size = b.get_u16().into();
+            if b.len() >= header_size {
+                //Break off the bytes after `header_size` and leave the header in `b`.
+                let rest = b.split_off(header_size);
```
Comment (giarc3) [inline src/v4/attached.rs]: I would keep using `split_to` and then the header can have a meaningful name

File: src/v5/mod.rs [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+            self.to_aes_value_with_attached_iv().as_ref(),
+        )
+    }
+
+    pub fn write_to_bytes(&self) -> Vec<u8> {
+        let mut result = Vec::with_capacity(self.0.len() + DETACHED_HEADER_LEN);
+        result.push(V0);
+        result.extend_from_slice(MAGIC);
+        result.extend_from_slice(self.0.as_ref());
+        result
```
Comment (giarc3) [inline src/v5/mod.rs:85]: Is this better than chaining?

File: src/v5/mod.rs [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+const MAGIC: &[u8; 4] = crate::v4::MAGIC;
+pub(crate) const V0: u8 = 0u8;
+/// This is 0 + IRON
+pub(crate) const DETACHED_HEADER_LEN: usize = 5;
+
+/// These are detached encrypted bytes, which means they have a `0IRON` + IV + CIPHERTEXT.
+/// This value is correct by construction and will be validated when we create it.
+/// There is no public constructor, only the TryFrom implementations.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct EncryptedPayload(pub IvAndCiphertext);
```
Comment (giarc3) [inline src/v5/mod.rs]: This inner type seems wrong, since it also has the 0IRON

**Review by giarc3 — CHANGES_REQUESTED** [review]
I reviewed it all, but it's still a jumble in my head. I'll keep reading through it to try to come up with further suggestions

File: src/v5/mod.rs (hunk 9 above)
Comment (coltfred) [reply src/v5/mod.rs:85]: It guarantees that we won't have to resize the vec at all. This seems better in the case that we know the exact length.

File: src/v4/attached.rs (hunk 8 above)
Comment (coltfred) [reply src/v4/attached.rs]: Sure. I don't have attachment either way.

**Review by giarc3 — APPROVED** [review]

**Pull request merged**